File: app/src/main/java/de/xikolo/controllers/main/MainFragment.java
package de.xikolo.controllers.main;

import de.xikolo.controllers.base.LoadingStatePresenterFragment;
import de.xikolo.presenters.base.LoadingStatePresenter;
import de.xikolo.presenters.base.LoadingStateView;

public abstract class MainFragment<P extends LoadingStatePresenter<V>, V extends LoadingStateView> extends LoadingStatePresenterFragment<P, V> {

    protected MainActivityCallback activityCallback;

    @Override
    public void onStart() {
        super.onStart();
        try {
            activityCallback = (MainActivityCallback) getActivity();
        } catch (ClassCastException e) {
            throw new ClassCastException(getActivity().toString() + " must implement MainActivityCallback");
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        activityCallback = null;
    }

    public interface MainActivityCallback {

        void onFragmentAttached(int id, String title);

        boolean isDrawerOpen();

        void updateDrawer();

        void selectDrawerSection(int pos);

    }

}

File: app/src/main/java/de/xikolo/controllers/main/MainFragment.kt
package de.xikolo.controllers.main

import android.os.Bundle
import android.view.View
import de.xikolo.App
import de.xikolo.extensions.observe
import de.xikolo.R
import de.xikolo.controllers.base.ViewModelFragment
import de.xikolo.viewmodels.base.BaseViewModel

abstract class MainFragment<T : BaseViewModel> : ViewModelFragment<T>() {

    var activityCallback: MainActivityCallback? = null

    override fun onStart() {
        super.onStart()
        try {
            activityCallback = activity as MainActivityCallback
        } catch (e: ClassCastException) {
            throw ClassCastException(activity.toString() + " must implement MainActivityCallback")
        }
    }

    override fun onStop() {
        super.onStop()
        activityCallback = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        App.instance.state.login
            .observe(viewLifecycleOwner) {
                onLoginStateChange(it)
            }
    }

    open fun onLoginStateChange(isLoggedIn: Boolean) {
        onRefresh()
    }

    protected fun showLoginRequired() {
        showLoginRequired {
            activityCallback?.selectDrawerSection(R.id.navigation_login)
        }
    }

}
